text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,165 @@
+   consumers of a message, regardless of intermediary instrumentation. The
+   creation context is created by the producer and must be propagated to the
+   consumers. It must not be altered by intermediaries.
+
+   This layer helps to model dependencies between producers and consumers,
+   regardless of the underlying messaging transport mechanism and its
+   instrumentation.
+2. The _transport context layer_ allows correlating the producer and the
+   consumer with an intermediary. It also allows to correlate multiple
+   intermediaries among each other. The transport context can be changed by
+   intermediaries, according to intermediary instrumentations. Intermediaries
+   that are not instrumented might simply drop the transport context.
+
+   This layer helps to gain insights into details of the message transport.
+
+A producer MUST attach a creation context to each message. The creation context
+MUST be attached in a way so that it is not changed by intermediaries. A
+producer MAY propagate a transport context to an intermediary.  An
+intermediary MAY propagate a transport context to a consumer.
+
+### Span structure, names, and attributes
+
+#### Consumer
+
+For many use cases, it is not possible to rely on the presence of "Process"
+spans for correlating producer with consumer traces: there are cases where a
+dedicated processing operation cannot be identified, or where processing
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Receive"
+spans to correlate producer with consumer traces. A "Receive" span must link
+to the "Create" spans of all messages that are handled by the respective
+"Receive" operation. Depending on the use case, "Receive" spans can correlate
+with "Process" spans or other spans modelling processing operations.
+
+The operation modelled by the "Receive" span does not strictly refer to
+receiving the message from intermediaries, but instead refers to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Receive" span. However, clients might pre-fetch messages
+from intermediaries and cache those messages, and only forward messages to the
+application at a later time. In this case, the operation of pre-fetching and
```

Comment:
> If the SDK implementation is receiving and delivering as a single operation, might it make sense to suggest it can skip the SDK Receive and implement only SDK Deliver?

Possibly, yes. My suggestion here is, that we don't put too many limitations on instrumenting the transport layer (the SDK Receive), but rather leave some freedom for SDK instrumentors to cover their use cases (which might differ a lot between SDKs), while still providing a common ground all those different SDK instrumentations can build on (e. g. the "Receive" span in the draft, we might rename it to avoid confusion).

I suggest keeping transport and application layer operations (receive and deliver) separate, but if SDK instrumentors feel compelled to cover both by the same span, I think the spec shouldn't prohibit that.

> This suggests Process could replace Deliver and Settle (where Status could carry ack/nack information). Just because this is possible, should this be done?

 I don't think so. I think we should strive for some communality between all different scenarios. I think a "Deliver" (or "Receive" span, as it's called in the draft) could be such a communality. Implementing "Process" spans is something we might suggest and recommend, but it's not something we can require.

Reply:
This should be covered by covering two scenarios (pull and push based) in the conventions.